apic_ml2/neutron/plugins/ml2/drivers/cisco/apic/mechanism_apic.py

```diff
@@ -1046,24 +1094,155 @@ def _perform_interface_port_operations(self, context, port, network,
         with self.apic_manager.apic.transaction() as trs:
             if is_vrf_per_router:
                 vrf_info = self._get_router_vrf(router)
+                router_name = router_id
+                if router.get('name'):
+                    router_name = router['name']
+                leak_ext_epg = self._get_leak_name_for_net(
+                    router_name, apic_manager.EXT_EPG)
                 if is_delete:
-                    # point BD back to the default VRF for this tenant
-                    # if router is not connected to any subnet anymore
-                    intf_ports = context._plugin.get_ports(
+                    if not network.get(cisco_apic.ALLOW_ROUTE_LEAK, False):
+                        route_leak_intf_nets = self._get_route_leak_networks(
+                            context, router_id)
+                        # delete the subnet of the external EPG for each route
+                        # leak network attached to this router
+                        intf_port_subnet = self._get_intf_port_subnet(context,
+                                                                      port)
+                        for route_leak_intf_net in route_leak_intf_nets:
+                            leak_l3out = self._get_leak_name_for_ext_net(
+                                route_leak_intf_net['name'])
+                            (self.apic_manager.
+                                ensure_external_epg_routes_deleted(
+                                    leak_l3out,
+                                    subnets=[intf_port_subnet['cidr']],
+                                    external_epg=leak_ext_epg,
+                                    owner=bd_tenant, transaction=trs))
+                    intf_net_ports = context._plugin.get_ports(
                         context._plugin_context,
                         filters={'device_owner':
                                  [n_constants.DEVICE_OWNER_ROUTER_INTF],
-                                 'device_id': [router['id']],
+                                 'device_id': [router_id],
                                  'network_id': [network['id']]})
-                    if not [p for p in intf_ports if p['id'] != port['id']]:
-                        vrf_default = self._get_network_vrf(context, network)
+                    if not [p for p in intf_net_ports
```

Comment:
this function has become too long, please move handling of the route-leak external-epg to separate functions that can be called from here.